### PR TITLE
Reorganize borrow check diagnostic code

### PR DESCRIPTION
**Pull request by mark-i-m**
Currently borrow checker diagnostics are split across many different modules in different places in the `librustc_mir` crate. This moves them all to a `diagnostics` module. This also reduces the nesting of the modules a bit (sooo much nesting).

I am also thinking of moving stuff out of the `nll` module since we only have one borrow checker now (:tada:), and maybe it even makes sense to split out all of this stuff to a `librustc_borrow_check`, but those are for the future. Feel free to ping me here or on zulip and let me know what you think...

cc @nikomatsakis @matthewjasper @eddyb

### CODE REVIEW

**Comment by mark-i-m** [comment]
I still need to fix up a bunch of imports... also, I'm not super happy about sprinkling `pub (in crate::borrow_check)` in front of the fields of `RegionInferenceContext`... thoughts?

**Comment by mark-i-m** [comment]
@matthewjasper Looks like this is ready for review. 

I would like to take a crack at making region_errors/region_name not be implemented in `RegionInferenceContext` at some point in a followup so that we could clean up the `pub(in crate::borrow_check)` messes...

**Comment by mark-i-m** [comment]
Sorry, one more thing: this will probably conflict with #66679, so maybe wait for that to merge first?

**Comment by eddyb** [comment]
I wouldn't mind `src/librustc_borrowck` (not `librustc_borrow_check` because both "rustc" and fully-spelled-out "check" seem a bit excessive, but that's just a stylistic prefer), *but*:
That's only if there is no dependency between the rest of `librustc_mir` and borrowck (in either direction), i.e. if they are two independent components that only happen to be colocated.

**Comment by matthewjasper** [comment]
Unfortunately there is a dependency until polonius is fully implemented.

**Comment by mark-i-m** [comment]
Sorry, forgetting the command right now, but this is S-blocked on #66679 

I rebased over that PR...

File: src/librustc_mir/borrow_check/diagnostics/mod.rs [hunk 1]
```diff
@@ -17,6 +19,21 @@ use super::borrow_set::BorrowData;
...
+mod find_use;
+mod var_name;
+mod region_name;
+mod outlives_suggestion;
+
+crate mod conflict_errors;
+crate mod move_errors;
+crate mod mutability_errors;
+crate mod region_errors;
+crate mod explain_borrow;
```
Comment (matthewjasper) [inline src/librustc_mir/borrow_check/diagnostics/mod.rs]: Can you make these private and re-export anything that's not public enough?

Comment (mark-i-m) [reply src/librustc_mir/borrow_check/diagnostics/mod.rs]: Woah. Very good call. I thought there were a lot of things, but it turns out there was only one thing that needed to be reexported that wasn't already!

**Comment by mark-i-m** [comment]
@matthewjasper  rebased

**Pull request merged**